apps/balancer-tools/src/app/apr/(utils)/tokenApr.ts

```diff
@@ -285,3 +321,159 @@ const getImplementationAddress = async (
+      apr = tokenYield * (1 - pool.protocolYieldFeeCache || 0.5);
+    }
+  } else if (
+    pool.poolType === PoolTypeEnum.META_STABLE ||
+    pool.poolType?.includes("Gyro")
+  ) {
+    apr = tokenYield * (pool.protocolYieldFeeCache || 0.5);
+  }
+
+  apr = (apr || 0) * tokenWeight;
+
+  return apr;
+}
+
+async function getTokenWeight(
+  tokenAddress: string,
+  startAtTimestamp: number,
+  endAtTimestamp: number,
+  poolNetwork: string,
+  poolId: string,
+): Promise<number> {
+  const rangeData = await poolsWithCache.gql(poolNetwork).poolSnapshotInRange({
+    poolId,
+    timestamp: generateDateRange(startAtTimestamp, endAtTimestamp),
+  });
+
+  let poolData: Pool;
+
+  if (rangeData.poolSnapshots.length > 0) {
+    poolData = rangeData.poolSnapshots[0].pool;
+  } else {
+    const currentPoolData = await pools.gql(poolNetwork).Pool({ poolId });
+    if (!currentPoolData || !currentPoolData.pool) {
+      throw new Error("Pool data is not found or is invalid.");
+    }
+    poolData = currentPoolData.pool;
+  }
+
+  const poolTokenData = poolData.tokens?.filter((token) => {
+    return token.address !== poolData.address;
+  });
+
+  if (!Array.isArray(poolTokenData)) {
+    return 0;
+  }
+
+  const token = poolTokenData.find((t) => t.address === tokenAddress);
+  if (!token) {
+    // eslint-disable-next-line no-console
+    console.warn(`Token not found in pool for address: ${tokenAddress}`);
+    return 0;
+  }
+
+  if (poolData.poolType === PoolTypeEnum.WEIGHTED && token.weight) {
+    return parseFloat(token.weight);
+  }
+
+  const tokensPrices = await Promise.all(
+    poolTokenData.map(async (token) => {
+      if (
+        vunerabilityAffecteRateProviders.some(
+          ({ address }) =>
+            address.toLowerCase() === token.address.toLowerCase(),
+        )
+      ) {
+        return 1;
+      } else {
+        const tokenPrice = await getTokenPriceByDate(
+          endAtTimestamp,
+          token.address,
+          parseInt(poolNetwork),
+        );
+
+        if (tokenPrice === undefined) {
+          // eslint-disable-next-line no-console
+          console.warn(
+            `Failed fetching price for ${token.symbol} (network: ${poolNetwork}, addr: ${token.address}) at ${endAtTimestamp}`,
+          );
+          return 1;
+        }
+
+        return tokenPrice;
+      }
+    }),
+  );
+
+  const totalValue = poolTokenData.reduce((acc, token, idx) => {
+    const balance = parseFloat(token.balance);
+    if (!isNaN(balance)) {
+      return acc + tokensPrices[idx] * balance;
+    }
+    return acc;
+  }, 0);
+
+  const tokenIdx = poolTokenData.findIndex((t) => t.address === tokenAddress);
+  const tokenPrice = tokensPrices[tokenIdx];
+
+  const tokenValue = tokenPrice * parseFloat(token.balance);
+  const tokenWeight = tokenValue / totalValue;
+
+  return tokenWeight;
+}
```

Comment:
again a bit hard to parse/understand. I'd aim for something like this:

```
async function getTokenWeight(
  tokenAddress: string, 
  startAtTimestamp: number, 
  endAtTimestamp: number, 
  poolNetwork: string, 
  poolId: string
): Promise<number> {
  const { tokens, poolType, address } = await fetchPoolData(poolNetwork, poolId, startAtTimestamp, endAtTimestamp);
  
  const relevantTokens = tokens?.filter(({ address }) => address !== address) || [];
  const targetToken = relevantTokens.find(({ address }) => address === tokenAddress);
  
  if (!targetToken) {
    console.warn(`Token not found: ${tokenAddress}`);
    return 0;
  }
  
  if (poolType === 'WEIGHTED' && targetToken.weight) return parseFloat(targetToken.weight);
  
  const tokenPrices = await fetchTokenPrices(relevantTokens, endAtTimestamp, poolNetwork);
  const totalValue = relevantTokens
    .map(({ balance }, idx) => parseFloat(balance) * tokenPrices[idx])
    .filter(val => !isNaN(val))
    .reduce((acc, val) => acc + val, 0);
  
  const tokenIdx = relevantTokens.findIndex(({ address }) => address === tokenAddress);
  return (parseFloat(targetToken.balance) * tokenPrices[tokenIdx]) / totalValue;
}
```

Reply:
also, since you're fetching the pool data from both functions, you could just fetch it beforehand in `getAPRFromRateProviderInterval` and pass it as an argument. Will save us another round trip to the subgraph =)